### ISSUE TITLE
260 bytes is maybe a bit much?

### ISSUE THREAD

**Issue by davidben** [tlswg/draft-ietf-tls-esni#134]
Stephen Farrell pointed out on the list that 260 is rather a lot of waste to stick into the ClientHello.
https://mailarchive.ietf.org/arch/msg/tls/kW_SWFamoj6IME4JzGseBKDyUV4

260 characters is this large of a hostname:
```
>>> "a" * 260
'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa'`
```

It's probably reasonable to recommend a smaller default value like 64 or 96 or 128. (Numbers pulled out of a hat. I have no data on how big it actually should be.)

**Comment by kazuho** [comment]
Yeah. The question is if we are happy with hostnames longer than the threshold (e.g., 64, 96, 128) sticking out. I think the current design of allowing the server to choose the correct value is the right approach.

OTOH, regarding the size you'd use for greasing, I can understand the concern of wasting bytes. Maybe browsers can use smaller sized payloads when greasing the ESNI extension, under the premise that they would be sometimes using the maximum.

The size of the extension being different from when it is actually used does not matter in practice, because an observer can typically tell if ESNI was used by checking if the unencrypted SNI record contains the fallback domain name (here I assume that we adopt #134).

**Comment by davidben** [comment]
Agreed that letting the server choose the correct value makes sense. But I think the specification should also be opinionated about this, so we don't get wildly different values for everyone, and so implementors have something to pick if they don't really care. The closely thing the spec currently has to an opinion is 260, which is kind of large.

**Comment by chris-wood** [comment]
Absent data that shows there's a tighter upper bound, I would prefer to keep the 260 recommendation and perhaps suggest that servers may choose something smaller if they know better.

**Comment by kontaxis** [comment]
According to RFC 1035 host names can be up to 255 octets.
I don't think this has changed.
So if this document makes a recommendation it should cover the worst case.

Someone could argue for something more efficient.
For instance, if we were to do draw the distribution of name lengths we would find that the median is 13 characters. (An excellent number) Min is 4 and max is 117. (Source: https://scans.io/study/umich-https)
But this can change and the recommendation in this document will become outdated.

Maybe add something like the following: "servers COULD tailor the padding length to cover the longest name associated with the hosts they handle." (And keep the existing number)

**Comment by bemasc** [comment]
In the current text, the "260" number only matters for GREASE; anyone who actually publishes ESNIKeys SHOULD choose the smallest number that covers their names, "rounded up to the nearest multiple of 16".  Given @kontaxis's data, the most common choices will likely be 16 and 32.  This state of affairs makes GREASE less effective, because (1) it doesn't resemble real usage and (2) it isn't exercising the length "joint".

As a solution, I would propose recommending that GREASE clients choose a valid length at random.  If we want to get fancy we could say that the random choice may be non-uniform (to favor shorter ClientHellos) but should have at least x% chance of choosing each length option.

Also, instead of saying that the padded_length "SHOULD" be 16*N or 260, I think we should consider a "MUST", and/or a denser encoding (e.g. enum with values 0..16).  Otherwise, we have to figure out what clients should do if `padded_length` is huge.  (I wonder how that interacts with QUIC and DDoS...)

**Comment by mrsylerpowers** [comment]
Another proposal is mentioned in #116 of just compressing these bits instead

**Comment by chris-wood** [comment]
@davidben this should be fixed, with new recommendation text, via #162. Please re-open if not!
